rfcs/referential-integrity.md

```diff
@@ -0,0 +1,156 @@
+2. We can't _just_ embed a reference to the record in the presentational layout, what if:
+  - The component or UI using the data doesn't adhere to the same interface, e.g, if it's existing or legacy code that can't be refactored for operational reasons?
+  - The data structure of the record changes, making the component that _did_ adhere to the data structure fall out of compliance?
+  
+Let's get more concrete. Imagine you had a design system with 50+ atomic elements (buttons, selects, text inputs, cards, etc). And above that, you already had your most common page building patterns in components (heros, call to actions, etc).
+
+To use these with the current discussed approach to referential integrity, the components would have to adhere to the model of an entity/content model. For example, this wouldn't work:
+
+```
+type Author {
+  id: number;
+  firstName: string;
+  lastName: string;
+}
+
+export const HeroBlock = {
+  component: ({data, index}) => (
+    <BlocksControls className="hero-block" index={index}>
+      <h1>{data.title}</h1>
+    </BlocksControls>
+  ),
+  template: {
+    id: "hero-block",
+    name: "Hero Block",
+    fields: [],
+    initialValues: {
+      title: "Enter a nice title"
+    }
+  }
+}
+```
+
+This won't work because author does not line up with the hero block's prop contract.
+
+### A Example Solution
+
+I believe the solution is by allowing editors to intuitively map a content model to the data structure of the block's props as the editor is building the reference:
+
+```
+import { join, firstLetter, resolveReferences } from "tinacms-reference";
+
+// Imagine we imported or read-in post, author, and layout from the filesystem
+const post = {
+  id: 1,
+  title: "Hello world",
+  description: "I'm alive",
+  author: 1
+}
+const author = {
+  id: 1,
+  firstName: "John",
+  lastName: "Doe"
+}
+const rowTemplate = {
+  component: (data) => (
+    <div>
+    </div>
+  ),
+  template: () => {
+    fields: 
+  }
+}
+const layout = {
+  id: 1,
+  blocks: [
+    {
+      _template: "hero",
+      title: {
+        entity: "page",
+        id: 1,
+        mapping: "title"
+      },
+      description: {
+        entity: "page",
+        id: 1,
+        mapping: "description"
+      },
+      author: {
+        entity: "author",
+        id: 1,
+        mapping: {
+          title: join(firstName, " ", lastName),
+          avatar: join(firstLetter(firstName), firstLetter(lastName))
+        }
+      }
+    }
+  ]
+}
+const form = useForm({
+  page: post,
+  author: author,
+  layout: resolveReferences(layout)
+}
+```
+
+## Value
+<!--
+  Please outline the value the work from this RFC would deliver to the users of TinaCMS.
+-->
+
+Making a workflow like the one above would allow relational data to live in Git, allowing us to treat relational data as code, and be much more intentional about it, and provide a lot more heuristics for understanding and manipulating structured content possible.
```

Comment:
I think it'd be helpful to know where this lives in the Tina ecosystem. As discussed yesterday we have a solution from our CMS graphql api which essentially treats file paths as foreign keys and gives us our relational tree (like we do in Forestry). Allowing a blog post to reference an author like so:

```md
---
title: 'Hello, World',
author: '/path-to-author.md'
---

Content
```

Are you saying this relational solution would be CMS-agnostic?

I'm also wondering why our library would export string utilities like `firstLetter` and `join` - does this give us some extra capabilities or is it for convenience?

Reply:
@jeffsee55 this are just pseudo code, as I'm imaging this being possible for the _editor_, not the developer, but don't have the UI abstraction in mind yet.